File: app/src/test/java/org/mifos/mobile/LoanAccountsTransactionPresenterTest.java
package org.mifos.mobile;

import android.content.Context;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mifos.mobile.api.DataManager;
import org.mifos.mobile.models.accounts.loan.LoanWithAssociations;
import org.mifos.mobile.presenters.LoanAccountsTransactionPresenter;
import org.mifos.mobile.ui.views.LoanAccountsTransactionView;
import org.mifos.mobile.util.RxSchedulersOverrideRule;
import org.mifos.mobile.utils.Constants;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;


import io.reactivex.Observable;

import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Created by dilpreet on 27/6/17.
 */
@RunWith(MockitoJUnitRunner.class)
public class LoanAccountsTransactionPresenterTest {

    @Rule
    public final RxSchedulersOverrideRule mOverrideSchedulersRule = new RxSchedulersOverrideRule();

    @Mock
    Context context;

    @Mock
    DataManager dataManager;

    @Mock
    LoanAccountsTransactionView view;

    private LoanAccountsTransactionPresenter presenter;
    private LoanWithAssociations loanWithTransaction, loanWithEmptyTransactions;

    @Before
    public void setUp() throws Exception {
        presenter = new LoanAccountsTransactionPresenter(dataManager, context);
        presenter.attachView(view);

        loanWithTransaction = FakeRemoteDataSource.getLoanAccountWithTransaction();
        loanWithEmptyTransactions = FakeRemoteDataSource.getLoanAccountWithEmptyTransaction();
    }

    @After
    public void tearDown() throws Exception {
        presenter.detachView();
    }

    @Test
    public void testLoadLoanAccountDetails() throws Exception {
        when(dataManager.getLoanWithAssociations(Constants.TRANSACTIONS, 29)).
                thenReturn(Observable.just(loanWithTransaction));

        presenter.loadLoanAccountDetails(29);

        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showLoanTransactions(loanWithTransaction);
        verify(view, never()).showEmptyTransactions(null);
        verify(view, never()).showErrorFetchingLoanAccountsDetail(context.
                getString(R.string.error_loan_account_details_loading));

    }

    @Test
    public void testLoadLoanAccountDetailsEmpty() throws Exception {
        when(dataManager.getLoanWithAssociations(Constants.TRANSACTIONS, 29)).
                thenReturn(Observable.just(loanWithEmptyTransactions));

        presenter.loadLoanAccountDetails(29);

        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showEmptyTransactions(loanWithEmptyTransactions);
        verify(view, never()).showLoanTransactions(null);
        verify(view, never()).showErrorFetchingLoanAccountsDetail(context.
                getString(R.string.error_loan_account_details_loading));

    }


    @Test
    public void testLoadLoanAccountDetailsFails() throws Exception {
        when(dataManager.getLoanWithAssociations(Constants.TRANSACTIONS, 29)).
                thenReturn(Observable.<LoanWithAssociations>error(new RuntimeException()));

        presenter.loadLoanAccountDetails(29);

        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showErrorFetchingLoanAccountsDetail(context.
                getString(R.string.error_loan_account_details_loading));
        verify(view, never()).showEmptyTransactions(null);
        verify(view, never()).showLoanTransactions(null);

    }
}

File: app/src/test/java/org/mifos/mobile/LoanAccountsTransactionPresenterTest.kt
package org.mifos.mobile

import android.content.Context

import io.reactivex.Observable

import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

import org.mifos.mobile.api.DataManager
import org.mifos.mobile.models.accounts.loan.LoanWithAssociations
import org.mifos.mobile.presenters.LoanAccountsTransactionPresenter
import org.mifos.mobile.ui.views.LoanAccountsTransactionView
import org.mifos.mobile.util.RxSchedulersOverrideRule
import org.mifos.mobile.utils.Constants

import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner

/**
 * Created by dilpreet on 27/6/17.
 */
@RunWith(MockitoJUnitRunner::class)
class LoanAccountsTransactionPresenterTest {
    @Rule
    @JvmField
    val mOverrideSchedulersRule = RxSchedulersOverrideRule()

    @Mock
    var context: Context? = null

    @Mock
    var dataManager: DataManager? = null

    @Mock
    var view: LoanAccountsTransactionView? = null
    private var presenter: LoanAccountsTransactionPresenter? = null
    private var loanWithTransaction: LoanWithAssociations? = null
    private var loanWithEmptyTransactions: LoanWithAssociations? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        presenter = LoanAccountsTransactionPresenter(dataManager!!, context)
        presenter?.attachView(view)
        loanWithTransaction = FakeRemoteDataSource.loanAccountWithTransaction
        loanWithEmptyTransactions = FakeRemoteDataSource.loanAccountWithEmptyTransaction
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        presenter?.detachView()
    }

    @Test
    @Throws(Exception::class)
    fun testLoadLoanAccountDetails() {
        Mockito.`when`(dataManager?.getLoanWithAssociations(Constants.TRANSACTIONS, 29)).thenReturn(Observable.just(loanWithTransaction))
        presenter?.loadLoanAccountDetails(29)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showLoanTransactions(loanWithTransaction)
        Mockito.verify(view, Mockito.never())?.showEmptyTransactions(null)
        Mockito.verify(view, Mockito.never())?.showErrorFetchingLoanAccountsDetail(context?.getString(R.string.error_loan_account_details_loading))
    }

    @Test
    @Throws(Exception::class)
    fun testLoadLoanAccountDetailsEmpty() {
        Mockito.`when`(dataManager?.getLoanWithAssociations(Constants.TRANSACTIONS, 29)).thenReturn(Observable.just(loanWithEmptyTransactions))
        presenter?.loadLoanAccountDetails(29)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showEmptyTransactions(loanWithEmptyTransactions)
        Mockito.verify(view, Mockito.never())?.showLoanTransactions(null)
        Mockito.verify(view, Mockito.never())?.showErrorFetchingLoanAccountsDetail(context?.getString(R.string.error_loan_account_details_loading))
    }

    @Test
    @Throws(Exception::class)
    fun testLoadLoanAccountDetailsFails() {
        Mockito.`when`(dataManager?.getLoanWithAssociations(Constants.TRANSACTIONS, 29)).thenReturn(Observable.error(RuntimeException()))
        presenter?.loadLoanAccountDetails(29)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showErrorFetchingLoanAccountsDetail(context?.getString(R.string.error_loan_account_details_loading))
        Mockito.verify(view, Mockito.never())?.showEmptyTransactions(null)
        Mockito.verify(view, Mockito.never())?.showLoanTransactions(null)
    }
}